File: ts/project/.sdk/tm/ts/src/feature/telemetry/TelemetryFeature.ts

import type { Context, FeatureOptions } from '../../types'
import type { ProjectNameSDK } from '../../ProjectNameSDK'

import { BaseFeature } from '../base/BaseFeature'


// Distributed-tracing telemetry. Opens a span per operation (PrePoint),
// propagates trace context to the server as W3C `traceparent` plus
// `X-Trace-Id` / `X-Span-Id` headers (PreRequest), and closes the span on
// completion (PreDone) or failure (PreUnexpected). Finished spans are kept
// on `client._telemetry.spans`; an `exporter` callback, when provided, is
// invoked with each finished span. Trace/span id generation and the clock
// are injectable for deterministic tests.
class TelemetryFeature extends BaseFeature {
  version = '0.0.1'
  name = 'telemetry'
  active = true

  _client?: ProjectNameSDK
  _options: any = {}
  _spans: WeakMap<object, any> = new WeakMap()
  _seq = 0


  init(ctx: Context, options: FeatureOptions): void | Promise<any> {
    this._client = ctx.client
    this._options = options || {}
    this.active = (options as any).active
    this._spans = new WeakMap()
    this._seq = 0

    const client: any = this._client
    if (null == client._telemetry) {
      client._telemetry = { spans: [], active: 0 }
    }
  }


  PrePoint(this: any, ctx: any) {
    const span = {
      traceId: this._id('trace'),
      spanId: this._id('span'),
      name: ((ctx.op && ctx.op.entity) || '_') + '.' + ((ctx.op && ctx.op.name) || '_'),
      start: this._now(),
      end: undefined as any,
      durationMs: undefined as any,
      ok: undefined as any,
    }
    this._spans.set(ctx, span)
    const client: any = this._client
    client._telemetry.active++
  }


  PreRequest(this: any, ctx: any) {
    const span = this._spans.get(ctx)
    const spec = ctx.spec
    if (null == span || null == spec) {
      return
    }
    if (null == spec.headers) {
      spec.headers = {}
    }
    const h = this._options.headers || {}
    spec.headers[h.trace || 'X-Trace-Id'] = span.traceId
    spec.headers[h.span || 'X-Span-Id'] = span.spanId
    spec.headers[h.parent || 'traceparent'] =
      '00-' + span.traceId + '-' + span.spanId + '-01'
  }


  PreDone(this: any, ctx: any) {
    this._close(ctx, !!(ctx.result && false !== ctx.result.ok && null == ctx.result.err))
  }


  PreUnexpected(this: any, ctx: any) {
    this._close(ctx, false)
  }


  _close(this: any, ctx: any, ok: boolean) {
    // Close once per operation; a PreDone followed by a pipeline throw
    // (non-2xx) fires PreUnexpected too, which then finds no open span.
    const span = this._spans.get(ctx)
    if (null == span) {
      return
    }
    this._spans.delete(ctx)
    span.end = this._now()
    span.durationMs = Math.max(0, span.end - span.start)
    span.ok = ok

    const client: any = this._client
    client._telemetry.active--
    client._telemetry.spans.push(span)

    const exporter = this._options.exporter
    if ('function' === typeof exporter) {
      try { exporter(span) } catch (_e) { }
    }
  }


  _id(this: any, kind: string): string {
    const idgen = this._options.idgen
    if ('function' === typeof idgen) {
      return idgen(kind)
    }
    // Deterministic-ish sequential id; unique within a client instance.
    this._seq++
    const n = this._seq.toString(16).padStart(4, '0')
    return (kind === 'trace' ? 't' : 's') + n.padEnd(16, '0')
  }


  _now(this: any): number {
    const now = this._options.now
    if ('function' === typeof now) {
      return now()
    }
    return Date.now()
  }
}


export {
  TelemetryFeature
}

File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/TelemetryFeature.swift
// Distributed-tracing telemetry. Opens a span per operation (prePoint),
// propagates trace context to the server as W3C `traceparent` plus
// `X-Trace-Id` / `X-Span-Id` headers (preRequest), and closes the span on
// completion (preDone) or failure (preUnexpected). Each span closes exactly
// once (the per-context marker in ctx.out is consumed on close). Finished
// spans accumulate on the feature; an `exporter` callback, when provided,
// is invoked with each finished span. Trace/span id generation (`idgen`)
// and the clock (`now`) are injectable for deterministic tests.

import Foundation

public final class TelemetryFeature: BaseFeature {
  private var client: ProjectNameSDK?
  private var options: VMap?
  private var seq = 0

  // Activity tracking (mirrors the ts client._telemetry record).
  public var spans: [VMap] = []
  public var activeSpans = 0

  private static let spanKey = "telemetry_span"

  public override init() {
    super.init()
    version = "0.0.1"
    name = "telemetry"
    active = true
  }

  public override func initFeature(_ ctx: Context, _ options: VMap) {
    client = ctx.client
    self.options = options
    active = foptBool(options, "active", false)
    seq = 0
  }

  public override func prePoint(_ ctx: Context) {
    if !active {
      return
    }

    let entity = ctx.op?.entity ?? "_"
    let opname = ctx.op?.name ?? "_"

    let span = VMap()
    span.entries["traceId"] = .string(id("trace"))
    span.entries["spanId"] = .string(id("span"))
    span.entries["name"] = .string(entity + "." + opname)
    span.entries["start"] = .int(foptNow(options)())
    ctx.out[TelemetryFeature.spanKey] = span
    activeSpans += 1
  }

  public override func preRequest(_ ctx: Context) {
    if !active {
      return
    }

    let span = ctx.out[TelemetryFeature.spanKey] as? VMap
    guard let span = span, let spec = ctx.spec else {
      return
    }

    let h = foptMap(options, "headers")
    let traceId = gp(span, "traceId").asString ?? ""
    let spanId = gp(span, "spanId").asString ?? ""
    spec.headers.entries[foptStr(h, "trace", "X-Trace-Id")] = .string(traceId)
    spec.headers.entries[foptStr(h, "span", "X-Span-Id")] = .string(spanId)
    spec.headers.entries[foptStr(h, "parent", "traceparent")] =
      .string("00-" + traceId + "-" + spanId + "-01")
  }

  public override func preDone(_ ctx: Context) {
    close(ctx, ctx.result != nil && ctx.result!.ok && ctx.result!.err == nil)
  }

  public override func preUnexpected(_ ctx: Context) {
    close(ctx, false)
  }

  private func close(_ ctx: Context, _ ok: Bool) {
    // Close once per operation; a preDone followed by a pipeline failure
    // (non-2xx) fires preUnexpected too, which then finds no open span.
    guard let span = ctx.out[TelemetryFeature.spanKey] as? VMap else {
      return
    }
    ctx.out.removeValue(forKey: TelemetryFeature.spanKey)

    let end = foptNow(options)()
    let start = toLong(gp(span, "start"))
    var dur = end - start
    if dur < 0 {
      dur = 0
    }
    span.entries["end"] = .int(end)
    span.entries["durationMs"] = .int(dur)
    span.entries["ok"] = .bool(ok)

    activeSpans -= 1
    spans.append(span)

    if let exporter = gp(options, "exporter").asNative as? (VMap) -> Void {
      exporter(span)
    }
  }

  private func id(_ kind: String) -> String {
    if let idgen = gp(options, "idgen").asNative as? (String) -> String {
      return idgen(kind)
    }
    // Deterministic-ish sequential id; unique within a client instance.
    seq += 1
    var n = String(seq, radix: 16)
    while n.count < 4 { n = "0" + n }
    let prefix = kind == "trace" ? "t" : "s"
    while n.count < 16 { n += "0" }
    return prefix + n
  }
}
